numpyro/distributions/directional.py

```diff
@@ -19,6 +19,143 @@
 )
 
 
+class SineSkewed(Distribution):
+    r"""Sine Skewing [1] is a procedure for producing a distribution that breaks pointwise symmetry on a torus
+    distribution. The new distribution is called the Sine Skewed X distribution, where X is the name of the (symmetric)
+    base distribution.
+
+    Torus distributions are distributions with support on products of circles
+    (i.e., :math:`\bigotimes^d S^1` where :math:`S^1=[-\pi, \pi)` ). So, a 0-torus is a point, the 1-torus is a circle,
+    and the 2-torus is commonly associated with the donut shape.
+
+    The Sine Skewed X distribution is parameterized by a weight parameter for each dimension of the event of X.
+    For example with a von Mises distribution over a circle (1-torus), the Sine Skewed von Mises Distribution has one
+    skew parameter. The skewness parameters can be inferred using :class:`~numpyro.infer.HMC` or
+    :class:`~numpyro.infer.NUTS`. For example, the following will produce a uniform prior over
+    skewness for the 2-torus::
+
+        def model(obs):
+            # Sine priors
+            phi_loc = numpyro.sample('phi_loc', VonMises(pi, 2.))
+            psi_loc = numpyro.sample('psi_loc', VonMises(-pi / 2, 2.))
+            phi_conc = numpyro.sample('phi_conc', Beta(2., 2.))
+            psi_conc = numpyro.sample('psi_conc', Beta(2., 2.))
+            corr_scale = numpyro.sample('corr_scale', Beta(2., 5.))
+
+            # SS prior
+            skew_phi = numpyro.sample('skew_phi', Uniform(-1., 1.))
+            psi_bound = 1 - skew_phi.abs()
+            skew_psi = numpyro.sample('skew_psi', Uniform(-1., 1.))
+            skewness = torch.stack((skew_phi, psi_bound * skew_psi), dim=-1)
+            with numpyro.plate('obs_plate'):
+                sine = SineBivariateVonMises(phi_loc=phi_loc, psi_loc=psi_loc,
+                                            phi_concentration=1000 * phi_conc,
+                                            psi_concentration=1000 * psi_conc,
+                                            weighted_correlation=corr_scale)
+                return numpyro.sample('phi_psi', SineSkewed(sine, skewness), obs=obs)
+
+    To ensure the skewing does not alter the normalization constant of the (Sine Bivaraite von Mises) base
+    distribution the skewness parameters are constraint. The constraint requires the sum of the absolute values of
+    skewness to be less than or equal to one.
+    So for the above snippet it must hold that::
+
+        skew_phi.abs()+skew_psi.abs() <= 1
+
+    We handle this in the prior by computing psi_bound and use it to scale skew_psi.
+    We do **not** use psi_bound as::
+
+        skew_psi = pyro.sample('skew_psi', Uniform(-psi_bound, psi_bound))
+
+    as it would make the support for the Uniform distribution dynamic.
```

Comment:
I'll add a PR for l1_ball to resolve this issue. In my opinion, this is quite inconvenient for users.

Reply:
Thanks, that would be nice!